### ISSUE TITLE
suboptimal watcher info transaction preparation 

### ISSUE THREAD

**Issue by InoMurko** [omisego/elixir-omg#1354]
https://github.com/omisego/elixir-omg/blob/v0.4.3/apps/omg_watcher_info/lib/omg_watcher_info/db/block.ex#L231-L240

http://erlang.org/doc/efficiency_guide/listHandling.html
```
{[tx | tx_list], outputs ++ output_list, inputs ++ input_list}
```
The following part can become *very very* unstable when blocks become full of transactions.
```
  @spec prepare_db_transactions(State.Transaction.Recovered.t(), pos_integer()) ::
          {[map()], [%DB.TxOutput{}], [%DB.TxOutput{}]}
  defp prepare_db_transactions(mined_transactions, block_number) do
    mined_transactions
    |> Stream.with_index()
    |> Enum.reduce({[], [], []}, fn {tx, txindex}, {tx_list, output_list, input_list} ->
      {tx, outputs, inputs} = prepare_db_transaction(tx, block_number, txindex)
IO.inspect output_list, label: "output_list"
IO.inspect outputs, label: "outputs"
IO.inspect inputs, label: "inputs"
IO.inspect input_list, label: "input_list"
      {[tx | tx_list], outputs ++ output_list, inputs ++ input_list}
    end)
  end
```

```
output_list: []
outputs: [
  %{
    amount: 1000000000000000000,
    blknum: 1000,
    creating_txhash: <<166, 13, 231, 202, 171, 116, 155, 201, 8, 20, 245, 164,
      236, 171, 229, 116, 22, 186, 107, 221, 156, 178, 113, 231, 126, 69, 112,
      55, 95, 165, 162, 55>>,
    currency: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>,
    oindex: 0,
    otype: 1,
    owner: <<235, 80, 193, 240, 71, 171, 44, 180, 88, 205, 43, 26, 12, 191, 62,
      186, 143, 63, 182, 49>>,
    txindex: 0
  },
  %{
    amount: 999999999999999999,
    blknum: 1000,
    creating_txhash: <<166, 13, 231, 202, 171, 116, 155, 201, 8, 20, 245, 164,
      236, 171, 229, 116, 22, 186, 107, 221, 156, 178, 113, 231, 126, 69, 112,
      55, 95, 165, 162, 55>>,
    currency: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>,
    oindex: 1,
    otype: 1,
    owner: <<114, 252, 118, 225, 233, 151, 86, 76, 104, 70, 183, 93, 247, 62,
      255, 159, 16, 167, 138, 178>>,
    txindex: 0
  }
]
inputs: [
  {{:utxo_position, 7, 0, 0}, 0,
   <<166, 13, 231, 202, 171, 116, 155, 201, 8, 20, 245, 164, 236, 171, 229, 116,
     22, 186, 107, 221, 156, 178, 113, 231, 126, 69, 112, 55, 95, 165, 162,
     55>>},
  {{:utxo_position, 6, 0, 0}, 1,
   <<166, 13, 231, 202, 171, 116, 155, 201, 8, 20, 245, 164, 236, 171, 229, 116,
     22, 186, 107, 221, 156, 178, 113, 231, 126, 69, 112, 55, 95, 165, 162,
     55>>}
]
input_list: []
output_list: [
  %{
    amount: 1000000000000000000,
    blknum: 1000,
    creating_txhash: <<166, 13, 231, 202, 171, 116, 155, 201, 8, 20, 245, 164,
      236, 171, 229, 116, 22, 186, 107, 221, 156, 178, 113, 231, 126, 69, 112,
      55, 95, 165, 162, 55>>,
    currency: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>,
    oindex: 0,
    otype: 1,
    owner: <<235, 80, 193, 240, 71, 171, 44, 180, 88, 205, 43, 26, 12, 191, 62,
      186, 143, 63, 182, 49>>,
    txindex: 0
  },
  %{
    amount: 999999999999999999,
    blknum: 1000,
    creating_txhash: <<166, 13, 231, 202, 171, 116, 155, 201, 8, 20, 245, 164,
      236, 171, 229, 116, 22, 186, 107, 221, 156, 178, 113, 231, 126, 69, 112,
      55, 95, 165, 162, 55>>,
    currency: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>,
    oindex: 1,
    otype: 1,
    owner: <<114, 252, 118, 225, 233, 151, 86, 76, 104, 70, 183, 93, 247, 62,
      255, 159, 16, 167, 138, 178>>,
    txindex: 0
  }
]
outputs: [
  %{
    amount: 1,
    blknum: 1000,
    creating_txhash: <<46, 145, 239, 174, 191, 196, 185, 252, 223, 213, 188, 93,
      140, 180, 98, 20, 198, 67, 169, 139, 104, 120, 111, 15, 82, 148, 141, 202,
      50, 32, 237, 178>>,
    currency: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>,
    oindex: 0,
    otype: 2,
    owner: <<59, 159, 76, 29, 210, 110, 11, 229, 147, 55, 59, 29, 54, 206, 226,
      0, 140, 190, 184, 55>>,
    txindex: 1
  }
]
inputs: []
input_list: [
  {{:utxo_position, 7, 0, 0}, 0,
   <<166, 13, 231, 202, 171, 116, 155, 201, 8, 20, 245, 164, 236, 171, 229, 116,
     22, 186, 107, 221, 156, 178, 113, 231, 126, 69, 112, 55, 95, 165, 162,
     55>>},
  {{:utxo_position, 6, 0, 0}, 1,
   <<166, 13, 231, 202, 171, 116, 155, 201, 8, 20, 245, 164, 236, 171, 229, 116,
     22, 186, 107, 221, 156, 178, 113, 231, 126, 69, 112, 55, 95, 165, 162,
     55>>}
]
```

**Comment by pdobacz** [comment]
I'm guessing you're referring to the `++` use specifically.

So I think the left-hand sides of the concatenation `++` are small in size (up to 4, usually 1 or 2) and taking on the advice in https://hexdocs.pm/elixir/Kernel.html#++/2

> The complexity of a ++ b is proportional to length(a), so avoid repeatedly appending to lists of arbitrary length, for example, list ++ [element]. Instead, consider prepending via [element | rest] and then reversing.

is it that much of a concern?

**Comment by InoMurko** [comment]
I actually went and measured it:

I've used two samples:
```
iex(50)> Enum.count mined_transactions
4000
iex(51)> Enum.count mined_transactions2
12000
``` 
where mined_transactions is:
```
txs = [
  %OMG.State.Transaction.Recovered{
    signed_tx: %OMG.State.Transaction.Signed{
      raw_tx: %OMG.State.Transaction.Payment{
        inputs: [{:utxo_position, 3005, 0, 0}, {:utxo_position, 3004, 0, 0}],
        metadata: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
          0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>,
        outputs: [
          %OMG.Output{
            amount: 1000000000000000000,
            currency: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
              0>>,
            output_type: 1,
            owner: <<148, 2, 179, 207, 58, 125, 33, 143, 44, 99, 34, 109, 191,
              165, 152, 149, 70, 171, 235, 157>>
          },
          %OMG.Output{
            amount: 999999999999999999,
            currency: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
              0>>,
            output_type: 1,
            owner: <<199, 86, 240, 27, 108, 61, 21, 90, 125, 45, 163, 45, 99,
              252, 196, 240, 29, 252, 190, 168>>
          }
        ],
        tx_type: 1
      },
      sigs: [
        <<51, 108, 142, 14, 40, 51, 246, 105, 182, 58, 53, 216, 95, 197, 54, 51,
          124, 73, 2, 206, 66, 248, 0, 46, 34, 150, 84, 76, 119, 237, 170, 198,
          103, 37, 142, 210, 221, 188, 54, 253, 203, 69, 55, 110, 212, 244, 242,
          72, 217, 243, 41, 248, 254, 114, 36, 167, 115, 247, 148, 196, 5, 96,
          139, 58, 27>>,
        <<51, 108, 142, 14, 40, 51, 246, 105, 182, 58, 53, 216, 95, 197, 54, 51,
          124, 73, 2, 206, 66, 248, 0, 46, 34, 150, 84, 76, 119, 237, 170, 198,
          103, 37, 142, 210, 221, 188, 54, 253, 203, 69, 55, 110, 212, 244, 242,
          72, 217, 243, 41, 248, 254, 114, 36, 167, 115, 247, 148, 196, 5, 96,
          139, 58, 27>>
      ]
    },
    signed_tx_bytes: <<249, 1, 93, 248, 134, 184, 65, 51, 108, 142, 14, 40, 51,
      246, 105, 182, 58, 53, 216, 95, 197, 54, 51, 124, 73, 2, 206, 66, 248, 0,
      46, 34, 150, 84, 76, 119, 237, 170, 198, 103, 37, 142, 210, 221, 188, 54,
      253, 203, 69, 55, 110, 212, 244, 242, 72, 217, 243, 41, 248, 254, 114, 36,
      167, 115, 247, 148, 196, 5, 96, 139, 58, 27, 184, 65, 51, 108, 142, 14,
      40, 51, 246, 105, 182, 58, 53, 216, 95, 197, 54, 51, 124, 73, 2, 206, 66,
      248, 0, 46, 34, 150, 84, 76, 119, 237, 170, 198, 103, 37, 142, 210, 221,
      188, 54, 253, 203, 69, 55, 110, 212, 244, 242, 72, 217, 243, 41, 248, 254,
      114, 36, 167, 115, 247, 148, 196, 5, 96, 139, 58, 27, 1, 248, 66, 160, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      2, 187, 167, 245, 34, 0, 160, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 187, 108, 90, 88, 0, 248, 108, 245, 1,
      243, 148, 148, 2, 179, 207, 58, 125, 33, 143, 44, 99, 34, 109, 191, 165,
      152, 149, 70, 171, 235, 157, 148, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 136, 13, 224, 182, 179, 167, 100, 0, 0, 245, 1, 243,
      148, 199, 86, 240, 27, 108, 61, 21, 90, 125, 45, 163, 45, 99, 252, 196,
      240, 29, 252, 190, 168, 148, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 136, 13, 224, 182, 179, 167, 99, 255, 255, 128, 160, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0>>,
    tx_hash: <<215, 74, 34, 238, 26, 208, 222, 195, 60, 103, 106, 145, 137, 77,
      186, 128, 88, 27, 229, 230, 198, 55, 49, 218, 102, 14, 14, 132, 181, 77,
      23, 20>>,
    witnesses: %{
      0 => <<199, 86, 240, 27, 108, 61, 21, 90, 125, 45, 163, 45, 99, 252, 196,
        240, 29, 252, 190, 168>>,
      1 => <<199, 86, 240, 27, 108, 61, 21, 90, 125, 45, 163, 45, 99, 252, 196,
        240, 29, 252, 190, 168>>
    }
  },
  %OMG.State.Transaction.Recovered{
    signed_tx: %OMG.State.Transaction.Signed{
      raw_tx: %OMG.State.Transaction.Fee{
        nonce: <<69, 173, 87, 49, 27, 127, 203, 232, 247, 200, 198, 181, 104,
          221, 89, 210, 210, 105, 76, 57, 67, 50, 252, 192, 202, 56, 81, 155,
          55, 137, 68, 192>>,
        outputs: [
          %OMG.Output{
            amount: 1,
            currency: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
              0>>,
            output_type: 2,
            owner: <<59, 159, 76, 29, 210, 110, 11, 229, 147, 55, 59, 29, 54,
              206, 226, 0, 140, 190, 184, 55>>
          }
        ],
        tx_type: 3
      },
      sigs: []
    },
    signed_tx_bytes: <<248, 82, 192, 3, 238, 237, 2, 235, 148, 59, 159, 76, 29,
      210, 110, 11, 229, 147, 55, 59, 29, 54, 206, 226, 0, 140, 190, 184, 55,
      148, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 160,
      69, 173, 87, 49, 27, 127, 203, 232, 247, 200, 198, 181, 104, 221, 89, 210,
      210, 105, 76, 57, 67, 50, 252, 192, 202, 56, 81, 155, 55, 137, 68, 192>>,
    tx_hash: <<105, 92, 82, 196, 185, 214, 193, 195, 1, 15, 57, 1, 175, 156,
      244, 225, 10, 61, 133, 58, 64, 191, 116, 19, 96, 17, 35, 187, 190, 101,
      61, 10>>,
    witnesses: %{}
  }
]
mined_transactions = Enum.map(1..4000, fn i ->
case rem(i,2) do
 1 -> hd(txs)
 0 -> hd(tl(txs))
end
end)
```
```
mined_transactions2 = mined_transactions ++ mined_transactions ++ mined_transactions
Benchee.run(
  %{
    "Block2" => fn -> OMG.WatcherInfo.DB.Block2.prepare_db_transactions(mined_transactions2, 4000) end,
    "Block" => fn -> OMG.WatcherInfo.DB.Block.prepare_db_transactions(mined_transactions2, 4000) end
  },
time: 0,
  warmup: 0,
memory_time: 2
)
```

```
iex(48)> Benchee.run(
...(48)>   %{
...(48)>     "Block2" => fn -> OMG.WatcherInfo.DB.Block2.prepare_db_transactions(mined_transactions2, 4000) end,
...(48)>     "Block" => fn -> OMG.WatcherInfo.DB.Block.prepare_db_transactions(mined_transactions2, 4000) end
...(48)>   },
...(48)> time: 0,
...(48)>   warmup: 0,
...(48)> memory_time: 2
...(48)> )
Operating System: macOS
CPU Information: Intel(R) Core(TM) i7-6567U CPU @ 3.30GHz
Number of Available Cores: 4
Available memory: 16 GB
Elixir 1.8.2
Erlang 22.1.1

Benchmark suite executing with the following configuration:
warmup: 0 ns
time: 0 ns
memory time: 2 s
parallel: 1
inputs: none specified
Estimated total run time: 4 s

Benchmarking Block...
Benchmarking Block2...

Memory usage statistics:

Name      Memory usage
Block2        44.63 MB
Block         44.72 MB - 1.00x memory usage +0.0917 MB

**All measurements for memory usage were the same**
%Benchee.Suite{
  configuration: %Benchee.Configuration{
    after_each: nil,
    after_scenario: nil,
    assigns: %{},
    before_each: nil,
    before_scenario: nil,
    formatters: [Benchee.Formatters.Console],
    inputs: nil,
    load: false,
    measure_function_call_overhead: true,
    memory_time: 2.0e9,
    parallel: 1,
    percentiles: '2c',
    pre_check: false,
    print: %{benchmarking: true, configuration: true, fast_warning: true},
    save: false,
    time: 0.0,
    title: nil,
    unit_scaling: :best,
    warmup: 0.0
  },
  scenarios: [
    %Benchee.Scenario{
      after_each: nil,
      after_scenario: nil,
      before_each: nil,
      before_scenario: nil,
      function: #Function<21.126501267/0 in :erl_eval.expr/5>,
      input: :__no_input,
      input_name: :__no_input,
      job_name: "Block2",
      memory_usage_data: %Benchee.CollectionData{
        samples: [46799768, 46799768, 46799768, 46799768],
        statistics: %Benchee.Statistics{
          absolute_difference: nil,
          average: 46799768.0,
          ips: nil,
          maximum: 46799768,
          median: 46799768.0,
          minimum: 46799768,
          mode: 46799768,
          percentiles: %{50 => 46799768.0, 99 => 46799768.0},
          relative_less: nil,
          relative_more: nil,
          sample_size: 4,
          std_dev: 0.0,
          std_dev_ips: nil,
          std_dev_ratio: 0.0
        }
      },
      name: "Block2",
      run_time_data: %Benchee.CollectionData{
        samples: [],
        statistics: %Benchee.Statistics{
          absolute_difference: nil,
          average: nil,
          ips: nil,
          maximum: nil,
          median: nil,
          minimum: nil,
          mode: nil,
          percentiles: nil,
          relative_less: nil,
          relative_more: nil,
          sample_size: 0,
          std_dev: nil,
          std_dev_ips: nil,
          std_dev_ratio: nil
        }
      },
      tag: nil
    },
    %Benchee.Scenario{
      after_each: nil,
      after_scenario: nil,
      before_each: nil,
      before_scenario: nil,
      function: #Function<21.126501267/0 in :erl_eval.expr/5>,
      input: :__no_input,
      input_name: :__no_input,
      job_name: "Block",
      memory_usage_data: %Benchee.CollectionData{
        samples: [46895896, 46895896, 46895896, 46895896, 46895896],
        statistics: %Benchee.Statistics{
          absolute_difference: 96128.0,
          average: 46895896.0,
          ips: nil,
          maximum: 46895896,
          median: 46895896.0,
          minimum: 46895896,
          mode: 46895896,
          percentiles: %{50 => 46895896.0, 99 => 46895896.0},
          relative_less: 0.9979501831034425,
          relative_more: 1.0020540272763745,
          sample_size: 5,
          std_dev: 0.0,
          std_dev_ips: nil,
          std_dev_ratio: 0.0
        }
      },
      name: "Block",
      run_time_data: %Benchee.CollectionData{
        samples: [],
        statistics: %Benchee.Statistics{
          absolute_difference: nil,
          average: nil,
          ips: nil,
          maximum: nil,
          median: nil,
          minimum: nil,
          mode: nil,
          percentiles: nil,
          relative_less: nil,
          relative_more: nil,
          sample_size: 0,
          std_dev: nil,
          std_dev_ips: nil,
          std_dev_ratio: nil
        }
      },
      tag: nil
    }
  ],
  system: %{
    available_memory: "16 GB",
    cpu_speed: "Intel(R) Core(TM) i7-6567U CPU @ 3.30GHz",
    elixir: "1.8.2",
    erlang: "22.1.1",
    num_cores: 4,
    os: :macOS
  }
}
iex(49)>
```


TLDR: the memory footprint is just a little bit bigger in 
```
{[tx | tx_list], outputs ++ output_list, inputs ++ input_list}
```
comparing it against (even though the end result isn't correct I've ignored that on purpose)
```
{[tx | tx_list], [outputs | output_list], [inputs | input_list]}
```
